File: internal/api/downloads.go
package api

import (
	"context"
	"errors"
	"net/http"
	"strconv"

	"github.com/jlbyh2o/llamaman/internal/hf/download"
	"github.com/jlbyh2o/llamaman/internal/model"
	"github.com/jlbyh2o/llamaman/internal/store"
)

// The download endpoints of DESIGN section 3.8.
//
// `POST /downloads` is the clearest example in this API of section 3's
// long-action rule: it answers `202 {"job_id","subject","model_id"}` — never the
// download itself — because the `jobs`, `downloads`, `models` and `model_files`
// rows are written in ONE transaction (section 2.7), so the job IS the receipt
// and `GET /downloads/{id}` carries the detail. An `Idempotency-Key` replay
// inside its window returns THE SAME BODY with `200` (D65), which is what makes
// a double-clicked Download a replay rather than a `409`.
//
// The two refusals both carry their numbers rather than merely refusing:
// `409 download_exists` names the download already running, and
// `409 insufficient_disk` says how much is free and how much is needed. A guard
// that will not say why is a guard a user cannot get past.

// DownloadService is everything this layer needs from internal/hf/download. The
// consumer owns the interface (DESIGN section 1); *download.Service satisfies it.
type DownloadService interface {
	Create(ctx context.Context, req download.CreateRequest) (download.CreateResult, error)
	Get(ctx context.Context, id string) (download.View, error)
	List(ctx context.Context, f store.DownloadFilter) ([]download.View, error)
	Pause(ctx context.Context, id string) error
	Resume(ctx context.Context, id string) error
	Retry(ctx context.Context, id string) error
	Cancel(ctx context.Context, id string, keepPartial bool) error
	SetPriority(ctx context.Context, id string, priority int) error
}

// DownloadDTO is one row of `GET /api/v1/downloads`.
type DownloadDTO struct {
	ID      string `json:"id"`
	ModelID string `json:"model_id"`
	// RepoID, Revision and PrimaryFile come from the model row, so a list does
	// not need one request per download to say what is being downloaded.
	RepoID      string `json:"repo_id"`
	Revision    string `json:"revision"`
	PrimaryFile string `json:"primary_file"`

	State         string `json:"state"`
	Priority      int    `json:"priority"`
	IncludeMmproj bool   `json:"include_mmproj"`

	BytesTotal int64 `json:"bytes_total"`
	BytesDone  int64 `json:"bytes_done"`
	// BytesAtStart is the offset this run resumed from. It is on the wire
	// because it is what makes `speed_bps` and `eta_sec` checkable: a client can
	// see that 38 of 40 GB were already there and that the reported rate
	// describes the two that moved.
	BytesAtStart int64  `json:"bytes_at_start"`
	SpeedBPS     int64  `json:"speed_bps"`
	ETASec       *int64 `json:"eta_sec"`

	Attempts     int     `json:"attempts"`
	ErrorCode    *string `json:"error_code"`
	ErrorMessage *string `json:"error_message"`

	CreatedAt  string  `json:"created_at"`
	StartedAt  *string `json:"started_at"`
	FinishedAt *string `json:"finished_at"`

	// Files is the per-file progress of section 7.3. A sharded download shows
	// five bars, not one: "47%" of a five-shard model tells a user nothing about
	// which shard is stuck.
	Files []DownloadFileDTO `json:"files"`
}

// DownloadFileDTO is one file's line in a download.
type DownloadFileDTO struct {
	ID          string `json:"id"`
	ModelFileID string `json:"model_file_id"`
	ModelID     string `json:"model_id"`
	Filename    string `json:"filename"`
	ShardIndex  int    `json:"shard_index"`
	ShardTotal  int    `json:"shard_total"`
	State       string `json:"state"`
	BytesTotal  int64  `json:"bytes_total"`
	BytesDone   int64  `json:"bytes_done"`
	// Etag is the BLOB NAME — for an LFS object the sha256 hex. The HTTP
	// validator is deliberately NOT on the wire: it is an internal resume
	// detail, it means nothing to a client, and publishing it would invite one
	// to send it somewhere.
	Etag      *string `json:"etag"`
	Attempts  int     `json:"attempts"`
	LastError *string `json:"last_error"`
}

// CreateDownloadRequest is the body of `POST /api/v1/downloads`.
type CreateDownloadRequest struct {
	RepoID string `json:"repo_id"`
	// Revision may be a branch, a tag or a commit; omitted means `main`. What is
	// STORED is always the resolved commit — section 2.6 forbids a branch name
	// in `models.revision`, because `main` names a different tree next week.
	Revision string `json:"revision,omitempty"`
	// Files are paths inside the repository. Naming any file of a shard set
	// expands to the whole set, and a set the repository holds only part of is
	// refused (section 7.3).
	Files []string `json:"files"`
	// IncludeMmproj defaults to true when omitted, which is why it is a pointer:
	// a plain bool cannot tell "the client said false" from "the client said
	// nothing", and those are opposite answers here.
	IncludeMmproj *bool `json:"include_mmproj,omitempty"`
	// MmprojFile picks between several projectors, for the repository that has
	// more than one and where section 7.2's preference rule declined to guess.
	MmprojFile string `json:"mmproj_file,omitempty"`
	Kind       string `json:"kind,omitempty"`
	Priority   int    `json:"priority,omitempty"`
}

// CreateDownloadResponse is section 3.8's `202`: the job receipt plus the model
// id, so the UI can open the model page before a byte has moved (section 3.10a).
type CreateDownloadResponse struct {
	JobReceiptDTO
	ModelID string `json:"model_id"`
	// MmprojModelID is the projector's own `models` row, null when none was
	// included. It is a separate row because a projector is separately reusable
	// across quantizations (section 7.3).
	MmprojModelID *string `json:"mmproj_model_id"`
	DownloadID    string  `json:"download_id"`
	BytesTotal    int64   `json:"bytes_total"`
}

// PatchDownloadRequest is `PATCH /api/v1/downloads/{id}`: the queue reorder.
type PatchDownloadRequest struct {
	Priority int `json:"priority"`
}

func (a *API) downloadRoutes() []Route {
	return []Route{
		a.listDownloadsRoute(),
		a.createDownloadRoute(),
		a.getDownloadRoute(),
		a.patchDownloadRoute(),
		a.pauseDownloadRoute(),
		a.resumeDownloadRoute(),
		a.retryDownloadRoute(),
		a.cancelDownloadRoute(),
	}
}

func (a *API) listDownloadsRoute() Route {
	return Route{
		Method:      http.MethodGet,
		Pattern:     BasePath + "/downloads",
		Auth:        AuthSession,
		OperationID: "listDownloads",
		Summary:     "Downloads with their per-file progress.",
		Tag:         "downloads",
		Query: []QueryParam{{
			Name: "state", Enum: []string{"active", "all"},
			Description: "`active` is everything unfinished, `paused` included; `all` is the whole " +
				"table, which is the receipt for what has landed on this disk.",
		}},
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			svc, err := a.downloads()
			if err != nil {
				WriteError(w, r, a.log, err)
				return
			}
			f := store.DownloadFilter{ActiveOnly: r.URL.Query().Get("state") != "all"}
			views, err := svc.List(r.Context(), f)
			if err != nil {
				WriteError(w, r, a.log, err)
				return
			}
			items := make([]DownloadDTO, 0, len(views))
			for _, v := range views {
				items = append(items, downloadDTO(v))
			}
			if err := WriteJSON(w, http.StatusOK, NewList(items, len(items), nil)); err != nil {
				WriteError(w, r, a.log, err)
			}
		}),
		Success: Response{Status: http.StatusOK, Description: "The downloads.",
			Body: List[DownloadDTO]{}},
	}
}

func (a *API) createDownloadRoute() Route {
	return Route{
		Method:      http.MethodPost,
		Pattern:     BasePath + "/downloads",
		Auth:        AuthSession,
		OperationID: "createDownload",
		Summary: "Queue a model download. The jobs, downloads, models and model_files rows are " +
			"written in one transaction, so the job is the receipt.",
		Tag:         "downloads",
		Idempotent:  true,
		RequestBody: CreateDownloadRequest{},
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			svc, err := a.downloads()
			if err != nil {
				WriteError(w, r, a.log, err)
				return
			}
			var body CreateDownloadRequest
			if err := DecodeJSON(w, r, &body); err != nil {
				WriteError(w, r, a.log, err)
				return
			}
			includeMmproj := true
			if body.IncludeMmproj != nil {
				includeMmproj = *body.IncludeMmproj
			}

			res, err := svc.Create(r.Context(), download.CreateRequest{
				RepoID:        body.RepoID,
				Revision:      body.Revision,
				Files:         body.Files,
				IncludeMmproj: includeMmproj,
				MmprojFile:    body.MmprojFile,
				Kind:          model.ModelKind(body.Kind),
				Priority:      body.Priority,
				Idempotency:   idempotencyFor(r, body),
			})
			if err != nil {
				a.writeDownloadError(w, r, err)
				return
			}

			// D65: a replay inside the window answers `200` with the same body.
			// A double-clicked Download must read as neither an error nor a
			// second download.
			status := http.StatusAccepted
			if res.Job.Replayed {
				status = http.StatusOK
			}
			out := CreateDownloadResponse{
				JobReceiptDTO: downloadReceipt(res.Job),
				ModelID:       res.ModelID,
				DownloadID:    res.Download.ID,
				BytesTotal:    res.Download.BytesTotal,
			}
			if res.MmprojModelID != "" {
				id := res.MmprojModelID
				out.MmprojModelID = &id
			}
			if err := WriteJSON(w, status, out); err != nil {
				WriteError(w, r, a.log, err)
			}
		}),
		Success: Response{
			Status:      http.StatusAccepted,
			Description: "The download was queued. Watch `job_id`.",
			Body:        CreateDownloadResponse{},
		},
		Errors: []Response{
			{
				Status:      http.StatusOK,
				Description: "An `Idempotency-Key` replay inside its window: the original download.",
				Body:        CreateDownloadResponse{},
			},
			{
				Status: http.StatusConflict,
				Description: "This model is already being downloaded (`details.download_id` names " +
					"it), or the target filesystem cannot hold it (`details` carries the numbers).",
				Codes: []model.ErrorCode{
					model.CodeDownloadExists, download.CodeInsufficientDisk, model.CodeJobInFlight,
				},
			},
			{
				Status: http.StatusForbidden,
				Description: "The repository is gated. `details` carries `repo` and `request_url`; " +
					"access grants are browser-only on the Hub's side, so the UI links out.",
				Codes: []model.ErrorCode{download.CodeHFGated, download.CodeHFPrivate},
			},
			{
				Status: http.StatusUnprocessableEntity,
				Description: "The request names files this repository does not hold at this " +
					"revision, a shard set the repository holds only part of, more than one " +
					"quantization, or a projector choice that is ambiguous.",
				Codes: []model.ErrorCode{
					download.CodeFileNotInRepo, download.CodeShardSetIncomplete,
					download.CodeNoFilesSelected, download.CodeMultipleQuants,
					download.CodeMmprojAmbiguous,
				},
			},
			{
				Status:      http.StatusBadGateway,
				Description: "The Hugging Face Hub could not be reached.",
				Codes:       []model.ErrorCode{download.CodeHFUnreachable},
			},
			{
				Status:      http.StatusServiceUnavailable,
				Description: "This daemon has no primary cache root yet.",
				Codes:       []model.ErrorCode{download.CodeNoCacheRoot},
			},
		},
	}
}

func (a *API) getDownloadRoute() Route {
	return Route{
		Method:      http.MethodGet,
		Pattern:     BasePath + "/downloads/{id}",
		Auth:        AuthSession,
		OperationID: "getDownload",
		Summary:     "One download with its per-file progress.",
		Tag:         "downloads",
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			svc, err := a.downloads()
			if err != nil {
				WriteError(w, r, a.log, err)
				return
			}
			v, err := svc.Get(r.Context(), r.PathValue("id"))
			if err != nil {
				a.writeDownloadError(w, r, err)
				return
			}
			if err := WriteJSON(w, http.StatusOK, downloadDTO(v)); err != nil {
				WriteError(w, r, a.log, err)
			}
		}),
		Success: Response{Status: http.StatusOK, Description: "The download.", Body: DownloadDTO{}},
		Errors: []Response{{
			Status: http.StatusNotFound, Description: "No download has this id.",
			Codes: []model.ErrorCode{CodeNotFound},
		}},
	}
}

func (a *API) patchDownloadRoute() Route {
	return Route{
		Method:      http.MethodPatch,
		Pattern:     BasePath + "/downloads/{id}",
		Auth:        AuthSession,
		OperationID: "reorderDownload",
		Summary: "Change a download's queue priority. Moves the jobs row and the downloads row " +
			"together, because the pool leases on the former.",
		Tag:         "downloads",
		RequestBody: PatchDownloadRequest{},
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			svc, err := a.downloads()
			if err != nil {
				WriteError(w, r, a.log, err)
				return
			}
			var body PatchDownloadRequest
			if err := DecodeJSON(w, r, &body); err != nil {
				WriteError(w, r, a.log, err)
				return
			}
			if body.Priority < 0 {
				WriteError(w, r, a.log, BadRequest("priority must not be negative"))
				return
			}
			if err := svc.SetPriority(r.Context(), r.PathValue("id"), body.Priority); err != nil {
				a.writeDownloadError(w, r, err)
				return
			}
			v, err := svc.Get(r.Context(), r.PathValue("id"))
			if err != nil {
				a.writeDownloadError(w, r, err)
				return
			}
			if err := WriteJSON(w, http.StatusOK, downloadDTO(v)); err != nil {
				WriteError(w, r, a.log, err)
			}
		}),
		Success: Response{Status: http.StatusOK, Description: "The reordered download.",
			Body: DownloadDTO{}},
		Errors: []Response{{
			Status: http.StatusNotFound, Description: "No download has this id.",
			Codes: []model.ErrorCode{CodeNotFound},
		}},
	}
}

func (a *API) pauseDownloadRoute() Route {
	return a.downloadActionRoute("pause", "pauseDownload",
		"Pause a download. The job releases its lease, the running transfers unwind, and every "+
			"`.incomplete` file stands where it is for the resume to continue from.",
		func(svc DownloadService, r *http.Request, id string) error {
			return svc.Pause(r.Context(), id)
		})
}

func (a *API) resumeDownloadRoute() Route {
	return a.downloadActionRoute("resume", "resumeDownload",
		"Resume a paused download. It continues from the byte each file reached.",
		func(svc DownloadService, r *http.Request, id string) error {
			return svc.Resume(r.Context(), id)
		})
}

func (a *API) retryDownloadRoute() Route {
	return a.downloadActionRoute("retry", "retryDownload",
		"Run a failed or canceled download again, resuming from whatever is on disk.",
		func(svc DownloadService, r *http.Request, id string) error {
			return svc.Retry(r.Context(), id)
		})
}

// downloadActionRoute is the shared shape of pause, resume and retry: a `202`
// with no body, because each of them starts work the job reports on rather than
// producing a result of its own.
func (a *API) downloadActionRoute(verb, opID, summary string,
	apply func(DownloadService, *http.Request, string) error) Route {

	return Route{
		Method:      http.MethodPost,
		Pattern:     BasePath + "/downloads/{id}/" + verb,
		Auth:        AuthSession,
		OperationID: opID,
		Summary:     summary,
		Tag:         "downloads",
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			svc, err := a.downloads()
			if err != nil {
				WriteError(w, r, a.log, err)
				return
			}
			id := r.PathValue("id")
			if err := apply(svc, r, id); err != nil {
				a.writeDownloadError(w, r, err)
				return
			}
			v, err := svc.Get(r.Context(), id)
			if err != nil {
				a.writeDownloadError(w, r, err)
				return
			}
			if err := WriteJSON(w, http.StatusAccepted, downloadDTO(v)); err != nil {
				WriteError(w, r, a.log, err)
			}
		}),
		Success: Response{Status: http.StatusAccepted, Description: "The download after the change.",
			Body: DownloadDTO{}},
		Errors: []Response{
			{Status: http.StatusNotFound, Description: "No download has this id.",
				Codes: []model.ErrorCode{CodeNotFound}},
			{Status: http.StatusConflict,
				Description: "The download's current state does not admit this change.",
				Codes:       []model.ErrorCode{download.CodeDownloadNotPausable}},
		},
	}
}

func (a *API) cancelDownloadRoute() Route {
	return Route{
		Method:      http.MethodPost,
		Pattern:     BasePath + "/downloads/{id}/cancel",
		Auth:        AuthSession,
		OperationID: "cancelDownload",
		Summary: "Cancel a download. The partial files are kept by default, so a retry resumes " +
			"rather than starting over.",
		Tag: "downloads",
		Query: []QueryParam{{
			Name: "keep_partial", Type: "boolean",
			Description: "Default true. False removes each `.incomplete` file as the transfer " +
				"that owns it releases its handle.",
		}},
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			svc, err := a.downloads()
			if err != nil {
				WriteError(w, r, a.log, err)
				return
			}
			// The default is true and the parameter is read as "false only when
			// the client said so". A cancel that silently discarded forty
			// gigabytes because a query parameter was misspelled would be the
			// worst possible reading of an ambiguous request.
			keep := true
			if v := r.URL.Query().Get("keep_partial"); v != "" {
				parsed, err := strconv.ParseBool(v)
				if err != nil {
					WriteError(w, r, a.log, BadRequest("keep_partial must be a boolean"))
					return
				}
				keep = parsed
			}
			id := r.PathValue("id")
			if err := svc.Cancel(r.Context(), id, keep); err != nil {
				a.writeDownloadError(w, r, err)
				return
			}
			v, err := svc.Get(r.Context(), id)
			if err != nil {
				a.writeDownloadError(w, r, err)
				return
			}
			if err := WriteJSON(w, http.StatusAccepted, downloadDTO(v)); err != nil {
				WriteError(w, r, a.log, err)
			}
		}),
		Success: Response{Status: http.StatusAccepted, Description: "The canceled download.",
			Body: DownloadDTO{}},
		Errors: []Response{
			{Status: http.StatusNotFound, Description: "No download has this id.",
				Codes: []model.ErrorCode{CodeNotFound}},
			{Status: http.StatusConflict, Description: "The download has already finished.",
				Codes: []model.ErrorCode{download.CodeDownloadNotPausable}},
		},
	}
}

// writeDownloadError maps this package's codes onto the statuses section 3.8
// pairs them with.
//
// It is a local mapping rather than an addition to statusForCode because the
// codes are internal/hf/download's own: they are declared beside the code that
// returns them, and the ONE layer that knows what a 403 means is this one. Every
// code it does not recognize falls through to WriteError, which is where a
// `not_found` from the store and every unclassified error already go.
func (a *API) writeDownloadError(w http.ResponseWriter, r *http.Request, err error) {
	var me model.Error
	if errors.As(err, &me) {
		if status, ok := downloadStatus[me.Code]; ok {
			WriteError(w, r, a.log, &Error{
				Status: status, Code: me.Code, Message: me.Message, Details: me.Details, Err: err,
			})
			return
		}
	}
	WriteError(w, r, a.log, err)
}

// downloadStatus is section 3.8's code-to-status table.
//
// `hf_gated` is a 403 and not a 404 deliberately: the repository exists, the
// user can see it in a browser, and telling them it does not exist would send
// them looking for a typo instead of clicking the link this response carries.
var downloadStatus = map[model.ErrorCode]int{
	download.CodeHFGated:             http.StatusForbidden,
	download.CodeHFPrivate:           http.StatusForbidden,
	download.CodeHFUnreachable:       http.StatusBadGateway,
	download.CodeInsufficientDisk:    http.StatusConflict,
	download.CodeShardSetIncomplete:  http.StatusUnprocessableEntity,
	download.CodeFileNotInRepo:       http.StatusUnprocessableEntity,
	download.CodeNoFilesSelected:     http.StatusUnprocessableEntity,
	download.CodeMultipleQuants:      http.StatusUnprocessableEntity,
	download.CodeMmprojAmbiguous:     http.StatusUnprocessableEntity,
	download.CodeDownloadNotPausable: http.StatusConflict,
	download.CodeNoCacheRoot:         http.StatusServiceUnavailable,
	model.CodeDownloadExists:         http.StatusConflict,
}

// downloadReceipt renders the `{job_id, subject}` receipt for a download job.
// It is a second function rather than a reuse of jobReceiptDTO because that one
// takes a models.JobRef, and converting between two identical structs at every
// call site would be worse than one four-line function.
func downloadReceipt(ref download.JobRef) JobReceiptDTO {
	out := JobReceiptDTO{
		Subject: SubjectDTO{Type: string(model.SubjectDownload), ID: ref.SubjectID},
	}
	if ref.JobID != "" {
		id := ref.JobID
		out.JobID = &id
	}
	return out
}

func downloadDTO(v download.View) DownloadDTO {
	d := DownloadDTO{
		ID:            v.ID,
		ModelID:       v.ModelID,
		RepoID:        v.RepoID,
		Revision:      v.Revision,
		PrimaryFile:   v.PrimaryFile,
		State:         string(v.State),
		Priority:      v.Priority,
		IncludeMmproj: v.IncludeMmproj,
		BytesTotal:    v.BytesTotal,
		BytesDone:     v.BytesDone,
		BytesAtStart:  v.BytesAtStart,
		SpeedBPS:      v.SpeedBPS,
		ETASec:        v.ETASec,
		Attempts:      v.Attempts,
		ErrorCode:     v.ErrorCode,
		ErrorMessage:  v.ErrorMessage,
		CreatedAt:     Time(v.CreatedAt),
		StartedAt:     TimePtr(v.StartedAt),
		FinishedAt:    TimePtr(v.FinishedAt),
		Files:         make([]DownloadFileDTO, 0, len(v.Tasks)),
	}
	for _, t := range v.Tasks {
		d.Files = append(d.Files, DownloadFileDTO{
			ID:          t.ID,
			ModelFileID: t.ModelFileID,
			ModelID:     t.ModelID,
			Filename:    t.Filename,
			ShardIndex:  t.ShardIndex,
			ShardTotal:  t.ShardTotal,
			State:       string(t.State),
			BytesTotal:  t.BytesTotal,
			BytesDone:   t.BytesDone,
			Etag:        t.Etag,
			Attempts:    t.Attempts,
			LastError:   t.LastError,
		})
	}
	return d
}

func (a *API) downloads() (DownloadService, error) {
	if a.cfg.Downloads == nil {
		return nil, Errorf(http.StatusServiceUnavailable, CodeInternalError,
			"this daemon was built without a download service")
	}
	return a.cfg.Downloads, nil
}

File: ui/src/features/models/downloads.ts
/**
 * The download queue — DESIGN section 3.8, section 7.4.
 *
 * A download is three rows written in one transaction (section 2.7): a `jobs` row that carries
 * scheduling, a `downloads` row that carries domain state, and one `download_tasks` row per file
 * folded upward into it. The UI only ever reads the middle one plus its files, because that fold is
 * the daemon's single writer and re-deriving it here would be a second, disagreeing opinion.
 *
 * Freshness comes from the `downloads` SSE topic: the worker pushes a patch every second, and
 * `applyFrame` merges it into both the detail query and every cached list. So there is no polling
 * here — `useLiveQueryOptions()` only starts an interval once the stream has actually given up.
 */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import type { UseQueryResult } from '@tanstack/react-query';

import { api } from '../../api/client';
import { ApiError } from '../../api/errors';
import { queryKeys } from '../../api/keys';
import type { Download, DownloadFile, ListPage } from '../../api/types';
import { useLiveQueryOptions } from '../../events/EventStreamProvider';
import { asPage } from './api';

/** `GET /api/v1/downloads` — `active` is everything unfinished, `all` is the whole ledger. */
export function useDownloads(
  state: 'active' | 'all' = 'active',
): UseQueryResult<ListPage<Download>, Error> {
  const live = useLiveQueryOptions(2_000);
  return useQuery({
    queryKey: queryKeys.downloads.list({ state }),
    queryFn: async () => asPage<Download>(await api.get('/api/v1/downloads', { query: { state } })),
    ...live,
  });
}

export interface CreateDownloadVars {
  repo_id: string;
  files: string[];
  revision?: string;
  include_mmproj: boolean;
  mmproj_file?: string;
  kind?: string;
  priority?: number;
}

/**
 * `POST /api/v1/downloads` — `202 {job_id, subject, model_id, download_id, bytes_total}`.
 *
 * The `Idempotency-Key` is not optional politeness here. This is the one button in the models area
 * that a double click would otherwise turn into two transfers of the same twenty gigabytes; with a
 * key, the replay inside the ten-minute window returns the original receipt with `200` (D65).
 */
export function useCreateDownload() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (vars: CreateDownloadVars) => {
      const { repo_id, files, revision, include_mmproj, mmproj_file, kind, priority } = vars;
      const body = {
        repo_id,
        files: [...files].sort(),
        include_mmproj,
        ...(revision === undefined ? {} : { revision }),
        ...(mmproj_file === undefined ? {} : { mmproj_file }),
        ...(kind === undefined ? {} : { kind }),
        ...(priority === undefined ? {} : { priority }),
      };
      return api.post('/api/v1/downloads', { body, idempotencyKey: downloadKey(body) });
    },
    onSuccess: async () => {
      await client.invalidateQueries({ queryKey: queryKeys.family('downloads') });
      await client.invalidateQueries({ queryKey: queryKeys.family('models') });
      await client.invalidateQueries({ queryKey: queryKeys.family('jobs') });
    },
  });
}

/**
 * A stable key for one request.
 *
 * It is derived from the **whole body**, not from a hand-picked subset, and that is the point:
 * section 3's rule is that the same key with a *different* body inside the ten-minute window is
 * `422 idempotency_key_reused`. Keying on "repository plus file set" alone would mean that starting
 * a download, canceling it, and restarting it at a different queue priority sent the same key with
 * a changed body — and the retry would be rejected as a replay of something it is not. Two clicks
 * on an unchanged dialog still produce one key, which is the collapse this exists for.
 */
export function downloadKey(body: Record<string, unknown>): string {
  // Key order must not affect the hash: the body is built with conditional spreads, so a projector
  // choice can move `mmproj_file` ahead of `kind` without the request having changed at all.
  const canonical = Object.keys(body)
    .sort()
    .map((key) => `${key}=${JSON.stringify(body[key])}`)
    .join('&');
  // The header must be a short printable ASCII token (section 3, `400` otherwise), so the body is
  // hashed rather than sent verbatim — a repo id and a file list are neither short nor ASCII-safe.
  return `dl-${hash32(canonical)}`;
}

/** FNV-1a, rendered hex. Not a security primitive: this only has to be stable and short. */
function hash32(input: string): string {
  let hash = 0x811c9dc5;
  for (let i = 0; i < input.length; i += 1) {
    hash ^= input.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193) >>> 0;
  }
  return hash.toString(16).padStart(8, '0');
}

type DownloadAction = 'pause' | 'resume' | 'retry';

/** `POST /api/v1/downloads/{id}/{pause|resume|retry}` — each answers `202` with the fresh row. */
export function useDownloadAction() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: async (vars: { id: string; action: DownloadAction }) => {
      const { id, action } = vars;
      if (action === 'pause') return api.post('/api/v1/downloads/{id}/pause', { path: { id } });
      if (action === 'resume') return api.post('/api/v1/downloads/{id}/resume', { path: { id } });
      return api.post('/api/v1/downloads/{id}/retry', { path: { id } });
    },
    onSuccess: async () => {
      await client.invalidateQueries({ queryKey: queryKeys.family('downloads') });
      await client.invalidateQueries({ queryKey: queryKeys.family('jobs') });
    },
  });
}

/**
 * `POST /api/v1/downloads/{id}/cancel?keep_partial=` — the default keeps the `.incomplete` files,
 * so a later retry resumes from where this stopped rather than starting the transfer again.
 */
export function useCancelDownload() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (vars: { id: string; keepPartial: boolean }) =>
      api.post('/api/v1/downloads/{id}/cancel', {
        path: { id: vars.id },
        query: { keep_partial: vars.keepPartial },
      }),
    onSuccess: async () => {
      await client.invalidateQueries({ queryKey: queryKeys.family('downloads') });
      await client.invalidateQueries({ queryKey: queryKeys.family('models') });
      await client.invalidateQueries({ queryKey: queryKeys.family('jobs') });
    },
  });
}

/**
 * `PATCH /api/v1/downloads/{id}` — queue order.
 *
 * Priority is ascending: the worker pool leases in `(priority, created_at)` order, so a *lower*
 * number runs sooner. The screen says "move up" and does the arithmetic, because nobody should have
 * to know that.
 */
export function useReorderDownload() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (vars: { id: string; priority: number }) =>
      api.patch('/api/v1/downloads/{id}', {
        path: { id: vars.id },
        body: { priority: vars.priority },
      }),
    onSuccess: async () => {
      await client.invalidateQueries({ queryKey: queryKeys.family('downloads') });
    },
  });
}

/* -- reading a download ---------------------------------------------------- */

/** States in which a transfer is still going somewhere. */
export const LIVE_DOWNLOAD_STATES: ReadonlySet<string> = new Set([
  'queued',
  'resolving',
  'running',
  'paused',
  'verifying',
]);

export function isLive(download: Download): boolean {
  return LIVE_DOWNLOAD_STATES.has(download.state);
}

/** Which of pause / resume / retry / cancel this state actually admits. */
export function downloadControls(state: string): {
  pause: boolean;
  resume: boolean;
  retry: boolean;
  cancel: boolean;
} {
  return {
    pause: state === 'running' || state === 'queued' || state === 'resolving',
    resume: state === 'paused',
    retry: state === 'failed' || state === 'canceled',
    cancel: LIVE_DOWNLOAD_STATES.has(state),
  };
}

export interface ShardRollup {
  /** The shard set's stem, or the filename itself when the file stands alone. */
  key: string;
  files: DownloadFile[];
  shardTotal: number;
  /** Files in `succeeded`, which is what "3 of 5 shards" counts. */
  done: number;
  bytesDone: number;
  bytesTotal: number;
}

/**
 * Group a download's files into shard sets.
 *
 * A sharded GGUF is one logical model (section 7.3) and the queue must read like one: five rows
 * named `…-00001-of-00005.gguf` are a rollup with a fraction, not five unrelated transfers. The
 * grouping is driven by `shard_total` from the DB rather than by re-parsing the filename, and the
 * stem is only used to *name* the group.
 */
export function rollupShards(files: readonly DownloadFile[]): ShardRollup[] {
  const groups = new Map<string, ShardRollup>();
  for (const file of files) {
    const key = file.shard_total > 1 ? shardStem(file.filename) : file.filename;
    let group = groups.get(key);
    if (!group) {
      group = {
        key,
        files: [],
        shardTotal: file.shard_total,
        done: 0,
        bytesDone: 0,
        bytesTotal: 0,
      };
      groups.set(key, group);
    }
    group.files.push(file);
    group.shardTotal = Math.max(group.shardTotal, file.shard_total);
    if (file.state === 'succeeded') group.done += 1;
    group.bytesDone += file.bytes_done;
    group.bytesTotal += file.bytes_total;
  }
  for (const group of groups.values()) {
    group.files.sort((a, b) => a.shard_index - b.shard_index);
  }
  return [...groups.values()];
}

/** `Qwen3-8B-Q4_K_M-00002-of-00005.gguf` → `Qwen3-8B-Q4_K_M`. */
export function shardStem(filename: string): string {
  const match = /^(.*?)-\d{5}-of-\d{5}\.gguf$/i.exec(filename);
  return match?.[1] ?? filename;
}

/** Percent complete, or null when the total is not known yet (`resolving`). */
export function downloadPercent(download: Download): number | null {
  if (download.bytes_total <= 0) return null;
  return Math.min(100, (download.bytes_done / download.bytes_total) * 100);
}

/**
 * The two `409`s of section 3.8, read off the error envelope.
 *
 * `download_exists` names the transfer already running, which the screen turns into a link rather
 * than an apology; `insufficient_disk` carries the numbers, which is the only form of that message
 * worth showing.
 */
export function downloadConflict(
  error: unknown,
): { kind: 'exists'; downloadId: string } | { kind: 'disk'; needed: number; free: number } | null {
  if (!(error instanceof ApiError)) return null;
  if (error.code === 'download_exists') {
    const id = error.details['download_id'];
    return { kind: 'exists', downloadId: typeof id === 'string' ? id : '' };
  }
  if (error.code === 'insufficient_disk') {
    const needed = error.details['needed_bytes'] ?? error.details['required_bytes'];
    const free = error.details['free_bytes'] ?? error.details['available_bytes'];
    return {
      kind: 'disk',
      needed: typeof needed === 'number' ? needed : 0,
      free: typeof free === 'number' ? free : 0,
    };
  }
  return null;
}
